### ISSUE TITLE
[FEATURE] Normalize data 

### ISSUE THREAD

**Issue by wissousad** [scikit-multiflow/scikit-multiflow#127]
Hello, I have a CSV file that has 9 features and 9 expected targets, and I want to test 2 regression models on this data (that should be generated as a stream).

When I test the `MultiTargetRegressionHoeffdingTree `  and `RegressorChain` on this data I get a bad R2-score, but when I tried normalizing my data with scikit-learn I get a pretty good R2-score. The problem is that I should not use scikit-learn because I should be normalizing data in real-time

What I really hope to see is a way to normalize data incrementally (in real-time) regardless of the model used. 

I tried using `normalize-samples(x) ` and `normalize-targets(y)`, but these methods depends on a parameter `samples-seen`  which is modified only during training. 

Here is my code: 
```
import csv
    from skmultiflow.data.file_stream import FileStream
    from skmultiflow.trees import MultiTargetRegressionHoeffdingTree
    from skmultiflow.evaluation import EvaluatePrequential
    from skmultiflow.meta import RegressorChain
    stream = FileStream('supervised_file_v4.csv', target_idx=9, n_targets=9)
    stream.prepare_for_use()
    ht = MultiTargetRegressionHoeffdingTree()
    rh = RegressorChain()
    ht.normalize_sample(stream.X)
    # here it gives me an array of 10 zeros beacause examples_seen = 0
    evaluator = EvaluatePrequential(show_plot=False,
                            pretrain_size=30000,
                            max_samples=200000, 
                           metrics = ['average_mean_absolute_error','average_root_mean_square_error'])
    # print(ht.examples_seen) gives 0
    evaluator.evaluate(stream=stream, model=[ht,rh], model_names=['HT', 'RH'])
    # print(ht.examples_seen) gives 200000
    ht.normalize_sample(stream.X)
    # even here it gives a error saying "operands could not be broadcast together with shapes (5733524,9) (9,) (5733524,) (9,)"
```

Any help or suggestion would be appreciated.

**Comment by smastelini** [comment]
Hello @wissousad,

The MultiTargetRegressionTree invokes the mentioned methods for normalizing your data (`normalize_sample` and `normalize_target`) for each new incoming sample. In other words, the tree model normalizes each incoming example before learning it. Thus, you do not need to call these methods externally.

Your score improves when all the data is used for normalization probably because all the properties of the dataset are observed a priori: mean, std, max, min (it depends on the type of normalization used).  On the other hand, the incremental algorithms only know an estimate of the underlying distribution in the dataset.

Given the aspects mentioned, I have some suggestions:

- Have you tried setting a different leaf predictor for the tree model? By default, MTR-HT uses `perceptron` as `leaf_predictor`. Very often, the usage of the `adaptive` model results in better results. `mean` is also another option.

- Moreover, are you calculating the R2 score externally? Your code is evaluating two well-known error metrics for Multi-target regression: average Mean Absolute Error (aMAE) and average Root Mean Square Error (aRMSE). I don't think R2 is currently implemented in skmultiflow for MTR tasks. [Borchani et al. (2015)](https://onlinelibrary.wiley.com/doi/full/10.1002/widm.1157) provide a really good overview of multi-output/multi-target regression. You can find more information on the mentioned evaluation metrics in their paper.

I am not much familiarized the `RegressionChain` implementation on skmultiflow. @jacobmontiel do you have any thoughts in that matter?

**Comment by wissousad** [comment]
Thank you so much @smastelini , I didn't know that the ` MultiTargetRegressionTree` does all that, That's great news. 
I calculate the R2_score using the `score ` method (from scikit-multiflow) and I also compare it to the one I get using sklearn, here's my code: 

```
tensor = list(stream.next_sample(100000)) 
x = tensor[0]
targets_p = tensor[1]
y_pred_ht= ht.predict(x)
y_pred_rh= rh.predict(x)
from sklearn.metrics import r2_score
print("r2 score with scikit-multiflow for ht", ht.score(x,targets_p))
print("r2 score with scikit-learn for ht", r2_score(targets_p,y_pred_ht))
print("r2 score with scikit-multiflow for rh", rh.score(x,targets_p))
print("r2 score with scikit-learn for rh", r2_score(targets_p,y_pred_rh))
```
I always get better results with regressor chain. And I also noticed that sometimes the r2-score using scikit-multiflow is really good (like 0.89) but with sklearn I get a negative value which is really bad, and when I see the results I notice that there is a big diffrence between the predicted value and the real one. So I guess that there is a little difference between the two metrics. 

Thank you so much for the paper I'm gonna read that and see what I can do. And for the suggestion on changing the leaf I'm gonna do that and I'll let you know. 

Thank you so much for you help.

**Comment by jacobmontiel** [comment]
Thanks for the complete discussion @smastelini.

`RegressorChain` does not perform explicitly data normalization, this is performed by the base estimator **if applicable**. For example, if you use `RegressorChain` with `RegressionHoeffdingTree` as the base estimator.

A `normalize` transformer could be paired with any model via a `pipeline`. Is this something worth considering for future development @smastelini @abifet @jmread?

**Comment by wissousad** [comment]
Thank you so much @jacobmontiel for your answer, I'm gonna try changing the base estimator.
Otherwise, I think that you have done a great job, but adding such a feature will only be beneficial. 

Thank you very much.
